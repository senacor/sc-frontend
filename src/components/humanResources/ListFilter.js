import React, { Component } from 'react';
import * as actions from '../../actions';
import { getSubFilter } from '../../reducers/selector';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Checkbox from '@material-ui/core/Checkbox/Checkbox';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import ListItem from '@material-ui/core/ListItem/ListItem';
import List from '@material-ui/core/List/List';
import Divider from '@material-ui/core/Divider/Divider';
import Icon from '@material-ui/core/Icon/Icon';
import {Button} from "@material-ui/core";

export class ListFilter extends Component {
  constructor(props) {
    super(props);

    const filter = Object.assign(
      {},
      { searchString: '', values: '' },
      this.props.filter
    );
    const checked =
      filter.values === '' ? Object.keys(props.content) : filter.values;
    this.state = {
      filter: filter,
      checked: checked,
      isAllSelected: checked.length === Object.keys(props.content).length
    };
  }

  isFilterSet = checked => {
    return !(checked.length === Object.keys(this.props.content).length ||
              checked.length === 0);
  };

  isFilterEmpty = checked => {
    return checked.length === 0;
  }

  handleToggleSelectAll = () => {
    const newAllSelect = !this.state.isAllSelected;
    let newChecked;

    if (newAllSelect === true) {
      newChecked = Object.keys(this.props.content);
    } else {
      newChecked = [];
    }
    this.setState({
      isAllSelected: newAllSelect,
      checked: newChecked
    });
  };

  handleToggle = value => () => {
    const { checked } = this.state;
    const currentIndex = checked.indexOf(value);
    const newChecked = [...checked];

    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }

    if (this.isFilterSet(newChecked)) {
      // something is set in filter, but not everything
      this.setState({
        checked: newChecked,
        isAllSelected: false
      });
    } else if (this.isFilterEmpty(newChecked)) {
      // nothing is set in filter
      console.log('new', newChecked)
      this.setState({
        checked: newChecked,
        isAllSelected: false
      });
    } else {
      // everything is set in filter
      this.setState({
        checked: Object.keys(this.props.content),
        isAllSelected: true
      });
    }
  };

  setFilter = () => {
    let searchString = `${this.props.filterBy}=`;
    searchString += this.state.checked
      .map(value => {
        return this.props.content[value];
      })
      .join(`&${this.props.filterBy}=`);

    let filter = {
      searchString,
      values: this.state.checked
    };
    let payload = {
      filterGroup: this.props.filterGroup,
      filterBy: this.props.filterBy,

      filter: filter
    };
    this.props.addFilter(payload);
  };

  showSelectAll = () => {
    return (
      <ListItem
        key={'selectAll'}
        dense
        button
        onClick={this.handleToggleSelectAll}
      >
        <Checkbox
          id={'selectAll'}
          checked={this.state.isAllSelected}
          color={'primary'}
          checkedIcon={<Icon>check</Icon>}
          icon={<Icon />}
        />
        <ListItemText primary={'selectAll'} />
      </ListItem>
    );
  };
  showContent = value => {
    return (
      <ListItem key={value} dense button onClick={this.handleToggle(value)}>
        <Checkbox
          id={value}
          checked={this.state.checked.indexOf(value) !== -1}
          color={'primary'}
          checkedIcon={<Icon>check</Icon>}
          icon={<Icon />}
        />
        <ListItemText primary={value} />
      </ListItem>
    );
  };

  render() {
    return (
      <List>
        {this.showSelectAll()}
        <Divider />
        {Object.keys(this.props.content).map(this.showContent, this)}
        <Divider />
        <Button onClick={this.setFilter}
                style={{backgroundColor: '#06A781', float: 'right'}}
        >
          OK
        </Button>
      </List>

    );
  }
}

ListFilter.propTypes = {
  filterGroup: PropTypes.string.isRequired,
  filterBy: PropTypes.string.isRequired,
  content: PropTypes.object.isRequired
};

export default connect(
  (state, props) => ({
    filter: getSubFilter(props.filterGroup, props.filterBy)(state)
  }),
  {
    addFilter: actions.addFilter,
    deleteFilter: actions.deleteFilter
  }
)(ListFilter);
