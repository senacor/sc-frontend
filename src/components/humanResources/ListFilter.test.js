import React from 'react';
import { ListFilter } from './ListFilter';
import { shallow } from 'enzyme';
import FILTER_GROUPS from './filterGroups';
import HR_ELEMENTS from './hrElements';

describe('ListFilter Component', () => {
  it('should match snapshot', () => {
    const mockAddFilter = jest.fn();
    const mockDeleteFilter = jest.fn();

    let component = shallow(
      <ListFilter
        content={{
          ja: 'yes',
          nein: 'no',
          vielleicht: 'maybe'
        }}
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.PR_OCCASION}
        filter={{
          searchString: '',
          values: ''
        }}
        addFilter={mockAddFilter}
        deleteFilter={mockDeleteFilter}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should init Checkboxes by store', () => {
    const mockAddFilter = jest.fn();
    const mockDeleteFilter = jest.fn();

    let component = shallow(
      <ListFilter
        content={{
          ja: 'yes',
          nein: 'no',
          vielleicht: 'maybe'
        }}
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.PR_OCCASION}
        filter={{
          searchString: `${HR_ELEMENTS.PR_OCCASION}=yes&${
            HR_ELEMENTS.PR_OCCASION
          }=no`,
          values: ['ja', 'nein']
        }}
        addFilter={mockAddFilter}
        deleteFilter={mockDeleteFilter}
      />
    );

    expect(component.find('[id="ja"]').props().checked).toEqual(true);
    expect(component.find('[id="nein"]').props().checked).toEqual(true);
    expect(component.find('[id="vielleicht"]').props().checked).toEqual(false);
    expect(component.find('[id="selectAll"]').props().checked).toEqual(false);
  });

  it('should update the Filter if one checkbox is clicked', () => {
    const mockAddFilter = jest.fn();
    const mockDeleteFilter = jest.fn();

    let component = shallow(
      <ListFilter
        content={{
          ja: 'yes',
          nein: 'no',
          vielleicht: 'maybe'
        }}
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.PR_OCCASION}
        filter={{
          searchString: '',
          values: ''
        }}
        addFilter={mockAddFilter}
        deleteFilter={mockDeleteFilter}
      />
    );

    component
      .find('[primary="vielleicht"]')
      .parent()
      .simulate('click');

    let payload = {
      filterGroup: FILTER_GROUPS.HR,
      filterBy: HR_ELEMENTS.PR_OCCASION,
      filter: {
        searchString: `${HR_ELEMENTS.PR_OCCASION}=yes&${
          HR_ELEMENTS.PR_OCCASION
        }=no`,
        values: ['ja', 'nein']
      }
    };

    expect(mockAddFilter).toHaveBeenCalledTimes(1);
    expect(mockAddFilter).toHaveBeenCalledWith(payload);

    expect(component.find('[id="vielleicht"]').props().checked).toEqual(false);

    expect(component.find('[id="selectAll"]').props().checked).toEqual(false);
  });

  it('should set all Checkboxes if the last is unchecked clicked', () => {
    const mockAddFilter = jest.fn();
    const mockDeleteFilter = jest.fn();

    let component = shallow(
      <ListFilter
        content={{
          ja: 'yes',
          nein: 'no',
          vielleicht: 'maybe'
        }}
        filterGroup={FILTER_GROUPS.HR}
        filterBy={HR_ELEMENTS.PR_OCCASION}
        filter={{
          searchString: '',
          values: ''
        }}
        addFilter={mockAddFilter}
        deleteFilter={mockDeleteFilter}
      />
    );

    component
      .find('[primary="ja"]')
      .parent()
      .simulate('click');
    component
      .find('[primary="vielleicht"]')
      .parent()
      .simulate('click');

    expect(component.find('[id="ja"]').props().checked).toEqual(false);
    expect(component.find('[id="vielleicht"]').props().checked).toEqual(false);

    component
      .find('[primary="nein"]')
      .parent()
      .simulate('click');

    expect(component.find('[id="selectAll"]').props().checked).toEqual(true);
    expect(component.find('[id="ja"]').props().checked).toEqual(true);

    expect(component.find('[id="vielleicht"]').props().checked).toEqual(true);
    expect(component.find('[id="nein"]').props().checked).toEqual(true);
  });
});
