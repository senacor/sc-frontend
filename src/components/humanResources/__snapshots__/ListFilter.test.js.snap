// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListFilter Component should match snapshot 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)
    button={true}
    dense={true}
    key="selectAll"
    onClick={[Function]}
  >
    <WithStyles(Checkbox)
      checked={true}
      checkedIcon={
        <WithStyles(Icon)>
          check
        </WithStyles(Icon)>
      }
      color="primary"
      icon={<WithStyles(Icon) />}
      id="selectAll"
    />
    <WithStyles(ListItemText)
      primary="selectAll"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    dense={true}
    key="ja"
    onClick={[Function]}
  >
    <WithStyles(Checkbox)
      checked={true}
      checkedIcon={
        <WithStyles(Icon)>
          check
        </WithStyles(Icon)>
      }
      color="primary"
      icon={<WithStyles(Icon) />}
      id="ja"
    />
    <WithStyles(ListItemText)
      primary="ja"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    dense={true}
    key="nein"
    onClick={[Function]}
  >
    <WithStyles(Checkbox)
      checked={true}
      checkedIcon={
        <WithStyles(Icon)>
          check
        </WithStyles(Icon)>
      }
      color="primary"
      icon={<WithStyles(Icon) />}
      id="nein"
    />
    <WithStyles(ListItemText)
      primary="nein"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    dense={true}
    key="vielleicht"
    onClick={[Function]}
  >
    <WithStyles(Checkbox)
      checked={true}
      checkedIcon={
        <WithStyles(Icon)>
          check
        </WithStyles(Icon)>
      }
      color="primary"
      icon={<WithStyles(Icon) />}
      id="vielleicht"
    />
    <WithStyles(ListItemText)
      primary="vielleicht"
    />
  </WithStyles(ListItem)>
</WithStyles(List)>
`;
